crates/accelerate/src/sparse_observable/mod.rs

```diff
@@ -633,6 +633,257 @@ impl SparseObservable {
+                    let new_indices = t1.indices.to_vec();
+                    let new_coeff = t1.coeff * Complex64::new(0.5, 0.0);
+                    (new_bits, new_indices, new_coeff)
+                }
+                (BitTerm::Left, BitTerm::Right, false) => {
+                    // c * '-' - c * '+' = -0.5c * 'Y'
+                    let mut new_bits = t1.bit_terms.to_vec();
+                    new_bits[mismatch_pos] = BitTerm::Y;
+                    let new_indices = t1.indices.to_vec();
+                    let new_coeff = t1.coeff * Complex64::new(-0.5, 0.0);
+                    (new_bits, new_indices, new_coeff)
+                }
+                _ => {
+                    return None;
+                }
+            };
+
+            Some(
+                SparseTerm::new(
+                    t1.num_qubits,
+                    new_coeff,
+                    new_bits.into(),
+                    new_indices.into(),
+                )
+                .unwrap(),
+            )
+        } else {
+            None
+        }
+    }
+
+    /// Greedily combine the terms in the observable.
+    ///
+    /// For example, a term ``1.5 * "X+IZ"`` can be combined with the term ``-1.5 * "X+ZZ"``,
+    /// producing the term ``3.0 * "X+1Z"``.
+    ///
+    /// Keeps the original ordering of terms as much as possible.
+    pub fn compress(&self, tol: f64) -> SparseObservable {
+        let mut terms: Vec<SparseTerm> = self.iter().map(|t| t.to_term()).collect();
+        let dummy_term =
+            SparseTerm::new(0, Complex64::new(0.0, 0.0), [].into(), [].into()).unwrap();
+        let mut another_iter: bool = true;
+
+        while another_iter {
+            another_iter = false;
+            let mut num_modified: usize = 0;
+            for i in 0..terms.len() {
+                if terms[i].coeff().norm_sqr() <= tol * tol {
+                    continue;
+                }
+                for j in i + 1..terms.len() {
+                    if terms[j].coeff().norm_sqr() <= tol * tol {
+                        continue;
+                    }
+                    // try to combine terms[i] and terms[j], storing the result in terms[i]
+                    if let Some(combined) = self.try_combine_terms(&terms[i], &terms[j], tol) {
+                        terms[i] = combined;
+                        terms[j] = dummy_term.clone();
+                        num_modified += 1;
+                    }
+                }
+            }
+            if num_modified > 0 {
+                another_iter = true;
+            }
+        }
+
+        let mut out = SparseObservable::zero(self.num_qubits);
+        for term in terms {
+            if term.coeff.norm_sqr() <= tol * tol {
+                continue;
+            }
+            out.coeffs.push(term.coeff);
+            out.bit_terms.extend_from_slice(&term.bit_terms);
+            out.indices.extend_from_slice(&term.indices);
+            out.boundaries.push(out.indices.len());
```

Comment:
You could just use 
```suggestion
            out.add_term(&term.view());
```
I think, instead of doing this manually 🙂

Reply:
Here also this is how it was already done in the existing code `pub fn canonicalize`.

Reply:
Done!